[WIP] "esy release" command

andreypopp opened the pull request:
This PR ports a release process from reason-cli as `esy release` command.

This is WIP for now. So I'll mention just the changes from the reason-cli for now:

* Releases are created using `esy release RELEASE_TYPE` command where `RELEASE_TYPE` could be one of `dev`, `pack` or `bin`.
* Releases are created inside `_release` directory: `_release/{dev,pack,bin-ARCH}` for each corresponding release type.
* There's no need to maintain two `package.json` files like in reason-cli (one for npm and one for esy) — instead we derive them automatically.
* Release configuration `"releasedBinaries"` is managed as `"esy.release.releasedBinaries"`.
* Releases are generated as npm packages with no dependencies (even for `dev`-type releases). For `dev` release type in `postinstall` step we install Esy "manually" (still via npm but in a separate prefix) and then use it to install Esy package.

TODO:
- [x] Refactor `esyrelease.js`: now I just copied it over from reason-cli, prettify it, flowify it & make resulted shell scripts more pretty
- [x] The top package should be also included in the release (right now it is simply ignored).
- [x] Tests — I plan to cover release functionality with e2e-tests
- [ ] Relocate artefacts into user store (we need to fix user store to include padding like we do in `esy release`)
- [ ] Rewrite wrapper scripts in OCaml (measure startup time, it should be faster than bash as @jordwalke reported)
- [ ] Investigate bugs with permissions

andreypopp committed:
Add `esy release` command.

This is a direct copy from [reason-cli/releaseUtils/release.js][1].

The only change for now is that we are making release not inside
`package/` directory but inside `_release/<release_tag>`.

Later commits will bring more improvements to the command.

[1]: https://github.com/reasonml/reason-cli/blob/9d2ac1/releaseUtils/release.js

andreypopp committed:
Make `esy release` derive npm & esy release package.json meta

Now instead of requiring maintaining two `package.json` files (one for
npm install and another for esy install/build) we try to derive both
from the original `package.json` file.

Also few general refactorings.

andreypopp committed:
Do not use `outdent` for `esyrelease.js`

jordwalke commented:
Sounds wonderful!

> Rewrite wrapper scripts in OCaml (measure startup time, it should be faster than bash as @jordwalke reported)

Also, this will allow them to be cross platform (windows). This is the only major blocker to creating windows releases of nearly any ocaml/reason app since when we distribute the final release, we need a wrapper script that can run on windows (ocamlrun really is a wonderful thing).

jordwalke commented on esy/src/bin/esy.js:
```diff
@@ -267,6 +269,53 @@ async function buildEjectCommand(sandboxPath) {
+  await pfs.unlink(tarFilename);
+
+  // Copy esyrelease.js executable over to release package.
+  const esyReleaseCommandOrigin = require.resolve('./esyrelease.js');
```
Does this need to be included in the released package so that we don't need to include a dependency on `esy` for packs?

andreypopp replied:
No, it's not needed as it isn't executed on client install at all. I just kept it that way (similar to `reason-cli`) before we reach sensible test coverage.

jordwalke commented on esy/src/lib/child_process.js:
```diff
@@ -0,0 +1,96 @@
+/**
```
Where is this file from? If not original? What's the license?

andreypopp replied:
That's from yarn source tree. We probably need some standard header for all Esy source files, right?

jordwalke replied:
We can add them at some point - if this is already from yarn, that's simple.

jordwalke reviewed (commented):
Feel free to commit into the master "esy" branch whenever you like. Shouldn't break anything, even if incomplete. We can even cut a release for esy, and then officially get reason-cli on top of `esy release` whenever you are ready.

andreypopp committed:
Refactor esyrelease.js script to be a library

Also prettify & flowify it. We don't need to copy it over to release
directory anymore, instead we use it as a library to generate
postinstall scripts.

andreypopp committed:
Remove spurious indents generated in release scripts

andreypopp committed:
Make release less verbose

Capture stdout/stderr and print it only if error has occured.

andreypopp committed:
Put entire source tree into rel/ subdir.

This is the correct way as we perform a build from it.

andreypopp committed:
Remove unused code in release

andreypopp committed:
Add `forRelease` config for sandbox.

We probably need to move shouldBePersisted property away from build
spec and decide in builder what to do with it. For now I need to make
sure we can continue covering release command with tests so I'll
procceed with the current solution.

andreypopp committed:
Add command-env and needed metadat for release to make builder

Added:
- `final-install-path-set.txt` contains a set of installation paths for
  the sandbox.
- `command-env` contains the environment which is used for running
  commands with sandbox environment (the diff with `eject-env` is that
  `command-env` contains the top level package's `bin/` in `$PATH`)

andreypopp committed:
Make use of command-env and final-install-path-set.txt in release

andreypopp committed:
Add initial tests for the release command

andreypopp committed:
Update esy/yarn.lock

jordwalke commented on esy/src/release.js:
```diff
@@ -193,19 +193,30 @@
+  type: ReleaseType,
+  version: string,
+  origin: string,
+  forGithubLFS: boolean,
```
BTW: I ended up not needing to support github large file storage. The trick was that I bundled up each individual prebuilt binary for each dependency into their own tar/gz each of which ends up being under the github limit.
- It was pretty hacky how I did that: In `sh`, I computed the final `PATH` and split it on `:` to find all the directories for all esy store transitive dependencies. This makes it so you can use the global store instead of creating a new empty build store every time you release (so slow).  I think `esy` should probably expose a library/command that will compute the set of transitive store directories for you without having to do hacky string manipulations like [I did](https://github.com/reasonml/reason-cli/blob/master/releaseUtils/release.js#L465)

andreypopp replied:
> I think esy should probably expose a library/command that will compute the set of transitive store directories

Already added in 83a1af4. It's to support `release` command at the moment, but I think we'd need a more structured db at some point.

andreypopp commented:
@jordwalke `eject-env` is more strict (it's for building) — a `$SHELL` is set to `env -i sh` which messes up things during normal usage (I remember we had a problem with `vim` executed under `eject-env`). Also in `eject-env` you don't see your `bin/` in `$PATH`.

jordwalke commented:
Okay makes sense. Strictness is good for building, and also while building you shouldn't see your own `bin/` output path.

andreypopp committed:
Add sync counterparts for lib/fs

This is useful for tests at least.

andreypopp committed:
More test cases for release

1. minimal project tests an empty project
2. with-dep-with-binary project tests the project with a dep

andreypopp committed:
Add more sync utils to lib/fs

andreypopp committed:
More test case for release: dev & pack types.

Also add more assertion regarding stdout. It's a little bit fragile for
npm... but we are ok that point, I think.

andreypopp committed:
Emit `command-env` parametrized over store location

andreypopp committed:
Update release test snapshots

andreypopp commented:
@jordwalke there's a "bin" defined with the package name as its name in [Reason CLI release script](https://github.com/reasonml/reason-cli/blob/9d2ac1e98374e72f632723c4a727ea9273404c32/releaseUtils/release.js#L634-L638) — I'm not sure why this is needed. Is that for convenience?

andreypopp committed:
Reset $ESY__SANDBOX variable when exec-ing prerelease.sh

andreypopp committed:
More robust tests

andreypopp commented:
Going to merge it as-is in `master` and iterate from there now on as suggested by @jordwalke.

I was able to produce releases for reason-cli and for esy-ocaml-project of all types though some modifications were needed to the reason-cli (mostly simplifications though).

andreypopp committed:
Add source type to build spec

I add `sourceType: 'transient' | 'immutable'` to `BuildSpec` type. This
is similar to `shouldBePersistent` but not quite.

It provides info if sources for the build are either transient (reside
in some local dir) or immutable (installed from release). Then builders
can decide on their own:

1. if they want to put artefacts in local or global store
2. if they need to rebuild sources on invokation

This commit also adds back the ability for makefile builder to clean &
rebuild sources for `transient` (previously `local`) builds. This is
more correct and now it doesn't hurt so much as we have simple builder
which can decide if it needs to rebuild intelligently (based on
checksums).

The motivation for this change was to have transient sources rebuilt for
releases. Next commit will update the tests cases.

andreypopp committed:
Update release test cases

jordwalke commented:
> @jordwalke there's a "bin" defined with the package name as its name in Reason CLI release script — I'm not sure why this is needed. Is that for convenience?

This was just to provide help for the entire package. However it recently bit me when I wanted to make the `esyvim` package export a binary named `esyvim` which was the way you would start `vim`. It collided with the package name. Perhaps it's not good to have this same-as-package-name automatically generated for each release.
